scripts/upsert_mgrb_participants.py

```diff
@@ -0,0 +1,177 @@
+            id
+            externalId
+            }
+        }
+    }
+    """
+)
+
+papi = ParticipantApi()
+
+
+def map_sample_eid_to_sample_data(data):
+    """Map sample external ID to sample data"""
+    return {sample['externalId']: sample for sample in data['sample']}
+
+
+def get_sample_data(project):
+    """Get sample data from metamist"""
+    response = query(SAMPLE_QUERY, {'project': project})
+    if not isinstance(response, dict):
+        raise ValueError('Expected a dictionary response from query')
+    return response
+
+
+def create_participant(sample_data: list, peid: str, sex: int):
+    """Create participant upsert object"""
+    participant_upsert = [
+        ParticipantUpsert(
+            id=None,
+            external_ids={'': peid},
+            reported_sex=sex,
+            samples=sample_data,
+        ),
+    ]
+    return participant_upsert
+
+
+def create_sample(project_id: int, seid: str, siid: str):
+    """Create sample upsert object"""
+    return SampleUpsert(
+        id=siid,
+        external_ids={'': seid},
+        project=project_id,
+    )
+
+
+def read_files_from_gcs(bucket_name, file_path):
+    """Read files from GCS"""
+    client = storage.Client()
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(file_path)
+    blob = blob.download_as_text(encoding='utf-8')
+    return StringIO(blob)
+
+
+@click.command()
+@click.option('--project', help='Project name to upsert participants to.')
+@click.option('--participant-meta', help='Path to participant metadata CSV.')
+@click.option('--sample-meta', help='Path to sample metadata CSV.')
+def main(project: str, participant_meta: str, sample_meta: str):
+    """Upsert participants to metamist"""
+    # Query metamist
+    data_response = get_sample_data(project)
+    project_id = data_response.get('project')['id']
+
+    sample_eid_mapping = map_sample_eid_to_sample_data(data_response)
+
+    # Set up logging
+    logging.basicConfig(level=logging.INFO)
+
+    # Read in csv
+    participant_data_io = read_files_from_gcs(BUCKET_NAME, participant_meta)
+    sample_data_io = read_files_from_gcs(BUCKET_NAME, sample_meta)
+
+    participant_data_list = list(csv.DictReader(participant_data_io))
+    sample_data_list = list(csv.DictReader(sample_data_io, delimiter='\t'))
+
+    # Count of samples for each participant
+    participant_sample_count: Dict[str, int] = defaultdict(int)
+    metadata_map = {}
+    for meta_row in participant_data_list:
+        metadata_map[meta_row['sampleID']] = meta_row
+        peid = meta_row['externalID']
+        participant_sample_count[peid] += 1
+
+    # Map peid to a list of sample upsert objects
+    participant_sex_map = {}
+    participant_sample_map: Dict[str, List[SampleUpsert]] = defaultdict(list)
+
+    for name_row in sample_data_list:
+        # Get the external ID
+        seid = '.'.join(name_row['CRAM'].split('.')[:2])
+
+        meta_row = metadata_map.get(name_row['RGSM (SampleID)'])
```

Comment:
If we want to generalise this script, we could pull this var out too.

Reply:
My experience with this fix has been that other scripts have tried to be generalised but there was nothing that quite fit this specific issue. I don't want to say this specific case won't happen again but it seems like a fairly niche scenario that we're trying to fix here especially given the makeup of metadata files from collaborators can be vastly different

Reply:
we might make 'RGSM (SampleID)' as input parameter to the script.